File: benchmarks/solutions/python/s1-channel-pipeline.py
from queue import Queue
from threading import Thread

def producer(out):
    for i in range(1, 21):
        out.put(i)
    out.put(None)

def filter_odd(inp, out):
    while (v := inp.get()) is not None:
        if v % 2 == 1:
            out.put(v)
    out.put(None)

def consumer(inp):
    while (v := inp.get()) is not None:
        print(v)

ch1, ch2 = Queue(), Queue()
Thread(target=producer, args=(ch1,)).start()
Thread(target=filter_odd, args=(ch1, ch2)).start()
consumer(ch2)

File: benchmarks/solutions/go/s1-channel-pipeline.go
package main

import "fmt"

func producer(out chan<- int) {
	for i := 1; i <= 20; i++ {
		out <- i
	}
	close(out)
}

func filterOdd(in <-chan int, out chan<- int) {
	for v := range in {
		if v%2 == 1 {
			out <- v
		}
	}
	close(out)
}

func main() {
	ch1 := make(chan int)
	ch2 := make(chan int)

	go producer(ch1)
	go filterOdd(ch1, ch2)

	for v := range ch2 {
		fmt.Println(v)
	}
}
